pkg/apis/kudo/v1alpha1/instance_types.go

```diff
@@ -16,65 +16,372 @@ limitations under the License.
+			}
+			return spec, nil
+		}
+	}
+	return nil, nil
+}
+
+// selectPlan returns nil if none of the plan exists, otherwise the first one in list that exists
+func selectPlan(possiblePlans []string, ov *OperatorVersion) *string {
+	for _, n := range possiblePlans {
+		if _, ok := ov.Spec.Plans[n]; ok {
+			return kudo.String(n)
+		}
+	}
+	return nil
+}
+
+// GetPlanToBeExecuted returns name of the plan that should be executed
+func (i *Instance) GetPlanToBeExecuted(ov *OperatorVersion) (*string, error) {
+	if i.GetPlanInProgress() != nil { // we're already running some plan
+		return nil, nil
+	}
+
+	// new instance, need to run deploy plan
+	if i.NoPlanEverExecuted() {
+		return kudo.String("deploy"), nil
+	}
+
+	// did the instance change so that we need to run deploy/upgrade/update plan?
+	instanceSnapshot, err := i.getSnapshotedSpec()
+	if err != nil {
+		return nil, err
+	}
+	if instanceSnapshot.OperatorVersion.Name != i.Spec.OperatorVersion.Name {
+		// this instance was upgraded to newer version
+		log.Printf("Instance: instance %s/%s was upgraded from %s to %s operatorversion", i.Namespace, i.Name, instanceSnapshot.OperatorVersion.Name, i.Spec.OperatorVersion.Name)
+		plan := selectPlan([]string{"upgrade", "update", "deploy"}, ov)
+		if plan == nil {
+			return nil, &InstanceError{fmt.Errorf("supposed to execute plan because instance %s/%s was upgraded but none of the deploy, upgrade, update plans found in linked operatorVersion", i.Namespace, i.Name), kudo.String("PlanNotFound")}
+		}
+		return plan, nil
+	}
+	if !reflect.DeepEqual(instanceSnapshot.Parameters, i.Spec.Parameters) {
+		// instance updated
+		log.Printf("Instance: instance %s/%s has updated parameters from %v to %v", i.Namespace, i.Name, instanceSnapshot.Parameters, i.Spec.Parameters)
+		paramDiff := parameterDifference(instanceSnapshot.Parameters, i.Spec.Parameters)
+		paramDefinitions := getParamDefinitions(paramDiff, ov)
+		plan := planNameFromParameters(paramDefinitions, ov)
+		if plan == nil {
+			return nil, &InstanceError{fmt.Errorf("supposed to execute plan because instance %s/%s was updatet but none of the deploy, update plans found in linked operatorVersion", i.Namespace, i.Name), kudo.String("PlanNotFound")}
+		}
+		return plan, nil
+	}
+	return nil, nil
+}
+
+// planNameFromParameters determines what plan to run based on params that changed and the related trigger plans
+func planNameFromParameters(params []Parameter, ov *OperatorVersion) *string {
+	for _, p := range params {
+		// TODO: if the params have different trigger plans, we always select first here which might not be ideal
+		if p.Trigger != "" && selectPlan([]string{p.Trigger}, ov) != nil {
+			return kudo.String(p.Trigger)
+		}
+	}
+	return selectPlan([]string{"update", "deploy"}, ov)
+}
+
+// getParamDefinitions retrieves parameter metadata from OperatorVersion CRD
+func getParamDefinitions(params map[string]string, ov *OperatorVersion) []Parameter {
+	defs := make([]Parameter, 0)
```

Comment:
super nit: (unless you insist on having size=0)
```suggestion
	defs := []Parameter{}
```